app/guided.py

```diff
@@ -0,0 +1,424 @@
+        self.goal_weight = goal_weight
+        self.fact_score_weight = fact_score_weight
+        self.last_action = None
+        self.goal_actions = set()
+
+        self.stopping_conditions = stopping_conditions
+        self.stopping_condition_met = False
+        self.state_machine = ['guided']
+        self.next_bucket = 'guided'
+
+    async def execute(self):
+        await self.planning_svc.execute_planner(self)
+    
+    async def guided(self):
+        await self.execute_subop(abilities=self.abilities, goals=self.goals)
+
+    async def execute_subop(self, abilities, agent=None, goals=[]):
+        """
+        :param abilities: List of ability IDs to plan with.
+        :param agent: Agent to focus planning on. If no agent is provided, planning will focus on all agents available in the operation.
+        :param goals: List of Goal objects to plan towards. Providing an empty list or an Exhaustion goal will cause the planner to infer goals from the ability's dependency graph.
+        """
+        attack_graph = await self._build_attack_graph(abilities, agent)
+
+        if len(goals) == 0 or goals[0].target == EXHAUSTION_KEY:
+            planner_goals = await self._create_terminal_goals(attack_graph)
+        else:
+            planner_goals = goals[:]
+
+        await self._show_attack_graph(attack_graph) # TODO: Remove before merging, just for debug purposes
+
+        absolute_distance_table = await self._build_distance_table(attack_graph, goals=planner_goals)
+        effective_distance_table = copy.deepcopy(absolute_distance_table)
+
+        goal_links = await self._get_goal_links(absolute_distance_table, agent=agent)
+        while len(planner_goals) > 0 and len(goal_links) > 0:
+            tasked_links = await self._task_agents(goal_links, effective_distance_table, abilities, agent=agent)
+
+            await self.operation.wait_for_links_completion(tasked_links)
+            
+            current_goal_count = len(planner_goals)
+            planner_goals = await self._update_goals(current_goals=planner_goals)
+            if len(planner_goals) < current_goal_count:
+                absolute_distance_table, effective_distance_table = await self._handle_goal_achieved(attack_graph, goals=planner_goals)
+            else:
+                effective_distance_table = await self._handle_no_goal_achieved(absolute_distance_table,
+                                                                                effective_distance_table,
+                                                                                goal_links)
+            goal_links = await self._get_goal_links(absolute_distance_table, agent=agent)
+
+        if not len(goal_links) or not len(planner_goals):
+            self.planning_svc.log.debug('No more links available or all goals satisfied, planner complete.')
+            self.next_bucket = None
+
+    """ PRIVATE """
+
+    async def _build_attack_graph(self, abilities, agent):
```

Comment:
In the context of adversaries with abilities that do not include input or output facts (the "check" adversary is one such example), the attack graph will be empty.  This seems to be working as intended, and more an issue with the adversary than with the planning or graph building logic.  It does, however, provide an opportunity to revert to standard `atomic` behavior in the case of an empty attack graph, such that the adversary executes its abilities rather than deeming the operation complete.  This guard rail may or may not prove useful, and might be out of scope for planner design.